### ISSUE TITLE
Does GNUStep's `objc_msgSend()` need function pointer cast?

### ISSUE THREAD

**Issue by takanuva** [gnustep/libobjc2#23]
Apple's "new" runtime needs `objc_msgSend()` to be cast to the appropriate function pointer type before calling it. This seems to be necessary as - as I understood it - the calling convention is different on some platforms with actual parameters and with variadic parameters.

What about GNUStep's (and GCC's)? Does it need to be cast as well? Also, if it doesn't _need_, does it _support_ it? (I'm asking it here for documentation purposes, as I could probably infer this information from the assembly code within a few... hours, days?)

I'm asking that because I'm currently working on a research compiler with support for Objective-C, which itself largely depends on libobjc. I want to simplify portability issues as much as I can.

**Comment by davidchisnall** [comment]
TL;DR: Yes, you need the cast and it has nothing to do with how `objc_msgSend` is implemented.

GCC's runtime doesn't implement `objc_msgSend`.  The requirement for a function pointer cast is dependent on the host ABI.  It's usually (though not universally) safe to skip the cast if your return value is an integer and you are passing no more than 4 arguments.

The `objc_msgSend` function simply forwards all of the arguments to the method.  The cast is so that the compiler knows to put the parameters into the registers and stack locations that the callee expects.  Without the cast, it will put them in with their current types (possibly missing a sign- or zero-extension or truncation that the callee expects) in the locations that the variadic calling convention expects.  C makes no guarantee that the variadic calling convention is the same as the non-variadic calling convention (on some architectures, for example, the variadic convention passes all of the variadic parameters on the stack, whereas the equivalent non-variadic call will put some of them in registers).

Note: This is not a new issue for Apple's 'new' runtime: the NeXT runtime also required this.
